R/outbreak_step.R

```diff
@@ -90,70 +87,55 @@ outbreak_step <- function(case_data = NULL, disp.iso = NULL, disp.com = NULL,
+    # cases whose parents are asymptomatic are automatically missed;
+    # will draw this for infector_asym == FALSE
+    missed = TRUE,
+    isolated = FALSE, new_cases = NA
+  )][,
     # draws a sample to see if this person is asymptomatic
-    asym = as.logical(rbinom(n = total_new_cases, 1, prob = prop.asym)),
-    # draws a sample to see if this person is traced
-    missed = as.logical(rbinom(n = total_new_cases, 1, prob = 1 - prop.ascertain)),
-    # sample from the incubation period for each new person
-    incubfn_sample = inc_samples,
-    isolated = FALSE,
-    new_cases = NA
-  )
-
-
-  prob_samples <- prob_samples[exposure < infector_iso_time][, # filter out new cases prevented by isolation
-                                             `:=`(# onset of new case is exposure + incubation period sample
-                                               onset = exposure + incubfn_sample)]
-
-
-  # cases whose parents are asymptomatic are automatically missed
-  prob_samples$missed[vect_isTRUE(prob_samples$infector_asym)] <- TRUE
-
-  # If you are asymptomatic, your isolation time is Inf
-  prob_samples[, isolated_time := ifelse(vect_isTRUE(asym), Inf,
-                                        # If you are not asymptomatic, but you are missed,
-                                        # you are isolated at your symptom onset
-                                        ifelse(vect_isTRUE(missed), onset + onset_to_isolation(1),
-                                               # If you are not asymptomatic and you are traced,
-                                               # you are isolated at max(onset,infector isolation time) # max(onset,infector_iso_time)
-                                               ifelse(!vect_isTRUE(rep(quarantine, total_new_cases)),
-                                                      pmin(onset + onset_to_isolation(1), pmax(onset, infector_iso_time)),
-                                                      infector_iso_time)))]
-
+    asym := runif(.N) < prop.asym
+  ][
+    exposure < infector_iso_time # keep only news cases that are pre-isolation
+  ][,
+    onset := exposure + incubation_period(.N) # onset of new case is exposure + incubation period sample
+  ]
+
+  # draw a sample for missing
+  prob_samples[infector_asym == FALSE, missed := runif(.N) > prop.ascertain]
+
+  prob_samples[, isolated_time := {
+    ref_time <- onset + onset_to_isolation(.N)
+    fcase(
+      # If asymptomatic, never isolated: time is Inf
+      asym == TRUE, Inf,
+      # If not asymptomatic, but are missed, isolated at your symptom onset
+      missed == TRUE, ref_time,
+      # if quarantine is in effect, isolated at infector's isolation time
+      !is.null(quarantine) && (quarantine == TRUE), infector_iso_time,
+      # isolated at symptom onset time if after infector isolation time,
+      # otherwise at the earlier of infector and infectee isolation times
+      default = pmin(ref_time, pmax(onset, infector_iso_time))
```

Comment:
so i think i've faithfully translated this, but one scientific aside: i'm unclear on what's going on here for quarantine.

seems like ... quarantine == FALSE would lead to earlier isolation times?

Reply:
you mean in case infectee isolation time is < infector isolation time? It seems in then they should get isolated at infectee isolation time but this is not happening here, i.e. this should be

```r
      !is.null(quarantine) && (quarantine == TRUE), pmin(ref_time, infector_iso_time),
```

Reply:
resolving as this is faithful to the previous implementation, can continue in #102